posthog/async_migrations/utils.py

```diff
@@ -66,6 +121,43 @@ def execute_op_postgres(sql: str, query_id: str):
         raise Exception(f"Failed to execute postgres op: sql={sql},\nquery_id={query_id},\nexception={str(e)}")
 
 
+def _get_number_running_on_cluster(query_pattern: str) -> int:
+    return sync_execute(
+        """
+        SELECT count()
+        FROM clusterAllReplicas(%(cluster)s, system, 'processes')
+        WHERE query LIKE %(query_pattern)s
+        """,
+        {"cluster": CLICKHOUSE_CLUSTER, "query_pattern": query_pattern},
+    )[0][0]
+
+
+def _sleep_until_finished(query_pattern: str) -> None:
+    from time import sleep
+
+    while _get_number_running_on_cluster(query_pattern) > 0:
+        logger.debug("Query still running, waiting until it's complete", query_pattern=query_pattern)
+        sleep(SLEEP_TIME_SECONDS)
+
+
+def run_optimize_table(unique_name: str, query_id: str, sql: str):
+    """
+    Runs the passed OPTIMIZE TABLE query.
+
+    Note that this handles process restarts gracefully: If the query is still running on the cluster,
+    we'll wait for that to complete first.
+    """
+    if _get_number_running_on_cluster(f"%%optimize:{unique_name}%%") > 0:
+        _sleep_until_finished(f"%%optimize:{unique_name}%%")
+    else:
+        execute_op_clickhouse(
+            f"/* optimize:{unique_name} */ " + sql,
```

Comment:
have you confirmed that this comment works? I believe later down the line we strip out all comments

Reply:
Yeah so `_prepare_query` will get rid of this